sarracen/interpolate.py

```diff
@@ -45,59 +45,42 @@ def interpolate2D(data: 'SarracenDataFrame',
 
     image = np.zeros((pixcounty, pixcountx))
 
-    # iterate through all pixels
-    for i, particle in data.iterrows():
-        # dimensionless weight
-        # w_i = m_i / (rho_i * (h_i) ** 2)
-        weight = particle['m'] / (particle['rho'] * particle['h'] ** 2)
-
-        # skip particles with 0 weight
-        if weight <= 0:
-            continue
-
-        # kernel radius scaled by the particle's 'h' value
-        radkern = kernel.radkernel * particle['h']
-        term = weight * particle[target]
-        hi1 = 1 / particle['h']
-        hi21 = hi1 ** 2
-
-        part_x = particle[x]
-        part_y = particle[y]
-
-        # determine the min/max x&y coordinates affected by this particle
-        ipixmin = int(np.rint((part_x - radkern - xmin) / pixwidthx))
-        jpixmin = int(np.rint((part_y - radkern - ymin) / pixwidthy))
-        ipixmax = int(np.rint((part_x + radkern - xmin) / pixwidthx))
-        jpixmax = int(np.rint((part_y + radkern - ymin) / pixwidthy))
-
-        # ensure that the min/max x&y coordinates remain within the bounds of the image
-        if ipixmin < 0:
-            ipixmin = 0
-        if ipixmax > pixcountx:
-            ipixmax = pixcountx
-        if jpixmin < 0:
-            jpixmin = 0
-        if jpixmax > pixcounty:
-            jpixmax = pixcounty
+    # clone the necessary data columns into a new dataframe for vectorized operations
+    parts = data.copy()[['m', 'rho', 'h', target, x, y]]
+    parts['weight'] = parts['m'] / (parts['rho'] * parts['h'] ** 2)
+
+    # filter out particles with 0 weight
+    parts = parts[parts['weight'] > 0]
+    parts['radkern'] = kernel.radkernel * parts['h']
+    parts['term'] = parts['weight'] * parts[target]
+    parts['hi21'] = (1 / parts['h']) ** 2
+
+    # determine maximum and minimum pixels that each particle contributes to
+    parts['ipixmin'] = np.rint((parts[x] - parts['radkern'] - xmin) / pixwidthx).clip(lower=0, upper=pixcountx)
+    parts['jpixmin'] = np.rint((parts[y] - parts['radkern'] - ymin) / pixwidthy).clip(lower=0, upper=pixcounty)
+    parts['ipixmax'] = np.rint((parts[x] + parts['radkern'] - xmin) / pixwidthx).clip(lower=0, upper=pixcountx)
+    parts['jpixmax'] = np.rint((parts[y] + parts['radkern'] - ymin) / pixwidthy).clip(lower=0, upper=pixcounty)
```

Comment:
See my comments in PR 6 about streamlining the memory requirements here + risk of 'target' duplication.